contracts/GameContract.sol

```diff
@@ -97,55 +147,76 @@ contract OneTwoThree {
     }
 
     // Send bet amount to create a room
-    function createRoom() public payable {
+    function createRoom( string memory _token) public payable {
         require(
             ownerCount[msg.sender] <= 10,
             "You have already reached to limit."
         );
-        require(msg.value > 0, "bet amount must greater than 0");
-
-        ownerCount[msg.sender] = ownerCount[msg.sender] + 1;
-        playerRoomsCount[msg.sender] = playerRoomsCount[msg.sender] + 1;
 
-        Room storage room = rooms[roomNumber];
-        room.id = roomNumber;
-        room.Address_1 = msg.sender;
-        room.Bet_amount = msg.value;
-        room.status = ROOM_STATUS.ONE_PERSON;
+        require( 
+            msg.value > 0, 
+            "bet amount must greater than 0"
+        );
 
-        emit CreateRoom(room);
+        if (sha256(bytes(_token)) != sha256(bytes("ETH"))){
+            require(
+                IERC20(list_token[_token]).allowance(msg.sender, address(this)) == msg.value,
+                "You must call approve first in web3"
+            );
+            require(
+                IERC20(list_token[_token]).transferFrom(msg.sender, address(this),msg.value),
+                "transfer failed"
+            );
+            ownerCount[msg.sender] = ownerCount[msg.sender] + 1;
+            playerRoomsCount[msg.sender] = playerRoomsCount[msg.sender] + 1;
+            rooms[roomNumber].status = ROOM_STATUS.ONE_PERSON;
+            emit CreateRoom(rooms[roomNumber]);
+            openingRoomNumber += 1;
+            roomNumber += 1;
+        } else {
+            rooms[roomNumber].status = ROOM_STATUS.ONE_PERSON;
```

Comment:
đoạn này trùng với đoạn trên điều kiện if thì sao k bỏ ra ngoài cho đỡ trùng code?

Reply:
Nếu bỏ ra ngoài thì điều kiện dù đúng hay sai thì nó đều được thực hiện như vậy dù chọn phòng không hợp kệ những biến đếm đó đều sẽ được tăng lên , làm ảnh hưởng đến những hàm khác

Reply:
Dùng require để chặn những case ko hợp lệ r mà, thằng nào hợp lệ thì mới tạo được room. Mà tạo room thì giống nhau với mọi loại token, k nên viết lại tận 6 dòng code trùng nhau

Reply:
vì mình sử dụng cả eth và erc20, nhưng eth không phải là erc20 nên các lệnh chuyển tiền của nó cũng khác với erc20. Vì vậy không thể sử dụng chung trong 1 require nên phải tách ra làm 2 .

Reply:
mình hiểu chỗ check cho erc20, nhưng check xong thì tạo room chung 1 chỗ cho cả eth và erc20 luôn
Code sẽ là kiểu như này:

`function createRoom( string memory _token) public payable {
        require(
            ownerCount[msg.sender] <= 10,
            "You have already reached to limit."
        );

        require( 
            msg.value > 0, 
            "bet amount must greater than 0"
        );

        if (sha256(bytes(_token)) != sha256(bytes("ETH"))){
            require(
                IERC20(list_token[_token]).allowance(msg.sender, address(this)) == msg.value,
                "You must call approve first in web3"
            );
            require(
                IERC20(list_token[_token]).transferFrom(msg.sender, address(this),msg.value),
                "transfer failed"
            );
        }
            rooms[roomNumber].status = ROOM_STATUS.ONE_PERSON;
            ownerCount[msg.sender] = ownerCount[msg.sender] + 1;
            playerRoomsCount[msg.sender] = playerRoomsCount[msg.sender] + 1;
            emit CreateRoom(rooms[roomNumber]);
            openingRoomNumber += 1;
            roomNumber += 1;
        rooms[roomNumber].Address_1 = msg.sender;
        rooms[roomNumber].Bet_amount = msg.value;
        rooms[roomNumber].token = _token;

    }`

Reply:
À chỗ check msg.value > 0 thì nên cho vào else của check token, vì erc20 thì k cần chuyển eth mà

Reply:
chỗ code theo mình nghĩ giả sử nó là erc20 nếu điều kiện require trong if (sha256(bytes(_token)) != sha256(bytes("ETH"))) không thỏa mãn thì nó sẽ chỉ không thực hiện lệnh trong if thôi còn lệnh ngoài if nó vẫn thực thi bình thường do đó nếu require không thỏa mãn thì nó vẫn tạo ra phòng mới

Reply:
> À chỗ check msg.value > 0 thì nên cho vào else của check token, vì erc20 thì k cần chuyển eth mà

cái điều kiện này là chung cho 2 loại erc20 và eth nên mình nghĩ nên để ngoài